### PR TITLE
Do not cancel context before reading body

### PR DESCRIPTION
**Pull request by michalpristas**
## What does this PR do?

As described in a related issue, sometimes it happens that context is cancelled before body is read. this then results in body being not fully processed and unmarshalling failing because Decoder is processing incomplete (trimmed) json body.

4-5 day long run with pprof should be performed before pushing this PR in

## Why is it important?

Fixes: https://github.com/elastic/kibana/issues/107514

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

File: x-pack/elastic-agent/pkg/remote/client.go [hunk 1]
```diff
@@ -165,15 +168,15 @@ func (c *Client) Send(
 	params url.Values,
 	headers http.Header,
 	body io.Reader,
-) (*http.Response, error) {
+) (*http.Response, CancelFunc, error) {
```
Comment (ruflin) [inline x-pack/elastic-agent/pkg/remote/client.go:171]: We should point out in the function comment that the caller is responsible for canceling it.

**Comment by ruflin** [comment]
Can you add a test that covers this issue, basically one that failed before but goes green now?

**Comment by ruflin** [comment]
Another question: Is there a way a user can work around this at the moment for example by changing a configuration?

**Comment by michalpristas** [comment]
@ruflin i'm not aware of any workaround or config option which would mitigate this.

**Review by scunningham — APPROVED** [review]
Looks correct to me.

**Comment by ruflin** [comment]
@michalpristas I suggest we get this in quickly and backport to 7.14. @EricDavisX Would be great if afterwards the team could run some manual tests with extra large policies.

**Comment by ruflin** [comment]
Can we please follow up with a test for this?

**Pull request merged**